### PR TITLE
Distinguish between seeds and manual peers

### PR DESCRIPTION
**Pull request by melekes**
Refs #864

### CODE REVIEW

**Comment by melekes** [comment]
- [ ] update changelog

**Comment by ebuchman** [comment]
Looks like a good start.

We should probably call them "persistent_peers"

Did you see the p2p writeup: https://docs.google.com/document/d/188tgD1t7D0oflZhCZGHuLUaFR3Caf6K_dB5WWgPqZPk/edit

Forgot to check in with that when I called it "manual" in the issues.

Also, I think we want this behaviour in the PEX reactor so that any time we run out we can fall back on seeds, not just the one time.

**Comment by melekes** [comment]
> Did you see the p2p writeup: 

no

**Comment by ebuchman** [comment]
I'll bring this home

File: node/node.go [hunk 1]
```diff
@@ -375,6 +382,27 @@ func (n *Node) OnStart() error {
...
+			outbound, inbound, _ := n.sw.NumPeers()
+			if n.IsRunning() && outbound+inbound == 0 {
+				// TODO: ignore error?
+				n.sw.DialPeersAsync(n.addrBook, seeds, false)
+			}
+		}()
+		return nil
+	}
+
+	return n.sw.DialPeersAsync(n.addrBook, seeds, false)
```
Comment (odeke-em) [inline node/node.go]: Minor nit, could we reverse the condition here to return early?
```go
if n.config.P2P.PexReactor && n.addrBook.Size() <= 0 {
   return n.sw.DialPeersAsync(n.addrBook, seeds, false)
}

// Give some time for PexReactor to connect us to other peers
```

File: p2p/pex_reactor_test.go [hunk 2]
```diff
@@ -118,16 +119,21 @@ func assertSomePeersWithTimeout(t *testing.T, switches []*Switch, checkPeriod, t
 					allGood = false
 				}
 			}
+			remaining -= checkPeriod
```
Comment (ebuchman) [inline p2p/pex_reactor_test.go]: why was this needed ?

File: p2p/pex_reactor_test.go [hunk 3]
```diff
@@ -180,6 +186,47 @@ func TestPEXReactorAbuseFromPeer(t *testing.T) {
 	assert.True(r.ReachedMaxMsgCountForPeer(peer.NodeInfo().ListenAddr))
 }
 
+func TestPEXReactorUsesSeedsIfNeeded(t *testing.T) {
+	dir, err := ioutil.TempDir("", "pex_reactor")
+	require.Nil(t, err)
+	defer os.RemoveAll(dir) // nolint: errcheck
+
+	book := NewAddrBook(dir+"addrbook.json", false)
```
Comment (ebuchman) [inline p2p/pex_reactor_test.go]: filepath.Join

File: p2p/switch.go [hunk 4]
```diff
@@ -315,15 +315,15 @@ func (sw *Switch) startInitPeer(peer *peer) {
...
+func (sw *Switch) DialPeersAsync(addrBook *AddrBook, peers []string, persistent bool) error {
+	netAddrs, errs := NewNetAddressStrings(peers)
 	for _, err := range errs {
-		sw.Logger.Error("Error in seed's address", "err", err)
+		sw.Logger.Error("Error in peer's address", "err", err)
 	}
 
 	if addrBook != nil {
-		// add seeds to `addrBook`
+		// add persistent peers to `addrBook`
```
Comment (ebuchman) [inline p2p/switch.go]: not necessarily persistent

File: rpc/core/net.go [hunk 5]
```diff
@@ -55,19 +54,31 @@ func NetInfo() (*ctypes.ResultNetInfo, error) {
...
 	logger.Info("DialSeeds", "addrBook", addrBook, "seeds", seeds)
-	err := p2pSwitch.DialSeeds(addrBook, seeds)
+	err := p2pSwitch.DialPeersAsync(addrBook, seeds, false)
 	if err != nil {
 		return &ctypes.ResultDialSeeds{}, err
 	}
 	return &ctypes.ResultDialSeeds{"Dialing seeds in progress. See /net_info for details"}, nil
 }
 
+func UnsafeDialPersistentPeers(persistent_peers []string) (*ctypes.ResultDialPersistentPeers, error) {
```
Comment (ebuchman) [inline rpc/core/net.go]: underscore ?

File: p2p/pex_reactor_test.go (hunk 3 above)
Comment (melekes) [reply p2p/pex_reactor_test.go]: no benefit. I believe `dir` already has `/` at the end

File: p2p/pex_reactor_test.go (hunk 2 above)
Comment (melekes) [reply p2p/pex_reactor_test.go]: because otherwise it will run forever. if you look closely you will see time.After() always resets in a loop
